core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/NativeImageBuildStep.java

```diff
@@ -81,6 +83,58 @@ ArtifactResultBuildItem result(NativeImageBuildItem image) {
         return new ArtifactResultBuildItem(image.getPath(), PackageConfig.NATIVE, Collections.emptyMap());
     }
 
+    @BuildStep(onlyIf = NativeSourcesBuild.class)
+    ArtifactResultBuildItem nativeSourcesResult(NativeConfig nativeConfig,
+            BuildSystemTargetBuildItem buildSystemTargetBuildItem,
+            NativeImageSourceJarBuildItem nativeImageSourceJarBuildItem,
+            OutputTargetBuildItem outputTargetBuildItem,
+            PackageConfig packageConfig,
+            List<NativeImageSystemPropertyBuildItem> nativeImageProperties) {
+
+        Path outputDir;
+        try {
+            outputDir = buildSystemTargetBuildItem.getOutputDirectory().resolve("native-sources");
+            IoUtils.createOrEmptyDir(outputDir);
+            IoUtils.copy(nativeImageSourceJarBuildItem.getPath().getParent(), outputDir);
+        } catch (IOException e) {
+            throw new UncheckedIOException("Unable to create native-sources output directory", e);
+        }
+
+        Path runnerJar = outputDir.resolve(nativeImageSourceJarBuildItem.getPath().getFileName());
+
+        final String runnerJarName = runnerJar.getFileName().toString();
+
+        CommandAndExecutable commandAndExecutable = new CommandAndExecutable.Builder()
+                .setNativeConfig(nativeConfig)
+                .setOutputTargetBuildItem(outputTargetBuildItem)
+                .setPackageConfig(packageConfig)
+                .setNativeImageProperties(nativeImageProperties)
+                .setOutputDir(outputDir)
+                .setRunnerJarName(runnerJarName)
+                // the path to native-image is not known now, it is only known at the time the native-sources will be consumed
+                .setNativeImageExecutable(Collections.emptyList())
+                // TODO: is this correct? It's set to true here because it's expected that native-sources will always be used
+                //  but it might make more sense to have some dedicated native-sources configuration
+                .setContainerBuild(true)
```

Comment:
Any particular reason why `isContainerBuild` couldn't be used here instead of hard-coding `true`?

Reply:
I'll add it